src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,489 @@
+        assert kvStoreInsertionPos > 0 && kvStoreInsertionPos < kvStore.length;
+        assert kvStoreInsertionPos % 2 == 0;
+
+        assert kvStoreInsertionPos >= hash.size; // because deletes only decrease hash.size
+        assert ((float) (2 * hash.size)) / index.length <= THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD;
+        return true;
+    }
+
+    private void insertIntoIndex(int keyHash, int kvPos) {
+        int pos = getIndexPosFromHash(keyHash, index.length);
+
+        boolean slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED_MARKER;
+        while (slotFilled) {
+            pos = (pos + 2) & (index.length - 1);
+            slotFilled = index[pos + 1] > INDEX_SLOT_UNUSED_MARKER;
+        }
+
+        index[pos] = keyHash;
+        index[pos + 1] = kvPos;
+    }
+
+    private void insertIntoKv(Object key, Object value) {
+        kvStore[kvStoreInsertionPos] = key;
+        kvStore[kvStoreInsertionPos + 1] = value;
+        kvStoreInsertionPos += 2;
+    }
+
+    private void resizeIndexIfNeeded(int size) {
+        if (size < numSlotsForIndexRebuild) {
+            return;
+        }
+
+        int[] oldIndex = this.index;
+        this.index = new int[2 * index.length];
+
+        for (int i = 0; i < oldIndex.length; i += 2) {
+            int hash = oldIndex[i];
+            int kvPos = oldIndex[i + 1];
+
+            insertIntoIndex(hash, kvPos);
+        }
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private void resizeKvStoreIfNeeded() {
+        if (kvStoreInsertionPos < kvStore.length) {
+            return;
+        }
+
+        kvStore = Arrays.copyOf(kvStore, 2 * kvStore.length);
+    }
+
+    private CompactHashStore copy() {
+        return new CompactHashStore(index.clone(), kvStore.clone(), kvStoreInsertionPos, numSlotsForIndexRebuild);
+    }
+
+    private Object deleteKvAndGetV(RubyHash hash, int hashPos, int keyKvPos) {
+        Object deletedValue = kvStore[keyKvPos + 1];
+        index[hashPos + 1] = INDEX_SLOT_DELETED_MARKER;
+        // TODO: Instead of naively nulling out the key-value, which can produce long gaps of nulls in the kvStore,
+        //       See if we can annotate each gap with its length so that iteration code can "jump" over it
+        kvStore[keyKvPos] = null;
+        kvStore[keyKvPos + 1] = null;
+
+        hash.size--;
+        return deletedValue;
+    }
+
+    /** @return the position of the first non-null key from the end of array, or a negative value if all keys are
+     *         null */
+    private int getLastKeyPos() {
+        int lastKeyPos = kvStoreInsertionPos - 2;
+        while (lastKeyPos > 0 && kvStore[lastKeyPos] == null) {
+            lastKeyPos -= 2;
+        }
+        return lastKeyPos;
+    }
+
+    private static int getIndexPosFromHash(int h, int max) {
+        int signMask = h >> 31;
+        // make positive
+        h = h ^ signMask;
```

Comment:
Would `h & ((1<<31)-1)` do the same? (if so it seems more efficient)
And if isn't the `h & (max - 1)` already enough on its own to make it positive?
Maybe we can even use `return h & (max - 2)` to do all of it at once? And `assert max >= 4 && isPowerOfTwo(4)`.
There is a definition of `isPowerOfTwo()` in IntegerNodes, you could move it to RubyGuards to reuse it.

I'm not sure we should discard the lowest bit of the hash though, it might be better to `(h & ((index.length>>1) - 1)) << 1` instead

Reply:
>Would h & ((1<<31)-1) do the same?

Hmmm, ((1 << 31) - 1) is the bitmask <0><1 {31 times}> right ? I had this idea too, although I thought of it more directly in terms of the mask shutting down the sign bit of the 2 comp representation. 

I rejected the idea at first because I thought it would be "less fair" than the shifted absolute value function I'm using, meaning that it results in some indices having more collisions than others. I don't remember my reasoning, but it was wrong! I checked now and that '&' actually have the same property as the absolute value, it just maps each negative numbers to exactly one positive number, just not its abs. Yup, checks out, it would work.

Reply:
> if isn't the h & (max - 1) already enough on its own to make it positive?

Also yes, thanks for pointing that out. Although I'm not sure that actually have the same semantics of "Make positive, THEN take the mod", but I don't think anybody cares about the mod operation per se, all we want is just a number between 0 and max-1, and h & (max - 1) does that nicely.

> return h & (max - 2)

Brilliant fusion of operations !

> assert max >= 4 && isPowerOfTwo(4)

I think you meant putting max inside the isPowerOfTwo call, not 4. Also, 2 is technically a valid power of 2 right ? it will always result in 0 which is the only valid index anyway. Regardless, max will always be >= 8 by the unstated precondition of the initialCapacity constructor, which is that the passed capacity be greater than or equal to the default initial capacity (8). I can assert that in the constructor itself.

Reply:
>I'm not sure we should discard the lowest bit of the hash though, it might be better to (h & ((index.length>>1) - 1)) << 1 instead

Hmm, can you elaborate further ? The current logic is not discarding the lowest bit of the hash, just the resulting mod (or mod-like) remainder. As far as I can see, that's the simplest way to ensure an even array index. What are some bad consequences of this in your opinion ?

Reply:
> I think you meant putting max inside the isPowerOfTwo call, not 4. Also, 2 is technically a valid power of 2 right ? it will always result in 0 which is the only valid index anyway. Regardless, max will always be >= 8 by the unstated precondition of the initialCapacity constructor, which is that the passed capacity be greater than or equal to the default initial capacity (8). I can assert that in the constructor itself.

Yes, it needs be a power of 2 >= 4, because if it's 2 then it would `& 0` and only cause collisions.
I would assert it here because this is the place it needs to hold. You can also assert it in other places of course, and e.g. create a helper method to asserts those 2 conditions.

> Hmm, can you elaborate further ? The current logic is not discarding the lowest bit of the hash, just the resulting mod (or mod-like) remainder. As far as I can see, that's the simplest way to ensure an even array index. What are some bad consequences of this in your opinion ?

It is keeping keeping the mod remainder, what is above is discarded. But if we do `return h & (max - 2)`, then we also discard the lowest bit, which seems not ideal.

Suppose I have some hash values like 6 and 7, if we discard the lowest bit, then those collide. But if we didn't discard it then they wouldn't collide.
We can't take all 32 bits of the hash, because our array will never have that much length, so I think the typical approach is to preserve all lower bits, because lower bits are most likely to differ for objects that are "close" in values to each other (e.g. if `int` hash is just itself).